### PR TITLE
feat: update project general info api

### PR DESCRIPTION
**Pull request by datphamcode295**
#### What's this PR do?

- [x] Update project's general info

#### Screenshots (if appropriate)
image

#### Request
```
curl --location --request PUT 'http://localhost:8200/api/v1/projects/8dc3be2e-19a4-4942-8a79-56db391a0b15/general-info' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM' \
--header 'Content-Type: application/json' \
--data-raw '{
    "name":"Fortress",
    "startDate":"1990-01-02T00:00:00+08:00",
    "countryID":"4ef64490-c906-4192-a7f9-d2221dadfe4c",
    "stacks":[
        "fa0f4e46-7eab-4e5c-9d31-30489e69fe2e",
        "b403ef95-4269-4830-bbb6-8e56e5ec0af4"
    ]
}'
```

#### Response
```
{
    "data": {
        "name": "Fortress",
        "startDate": "1990-01-02T00:00:00Z",
        "country": {
            "id": "4ef64490-c906-4192-a7f9-d2221dadfe4c",
            "name": "Vietnam",
            "code": "+84"
        },
        "stacks": [
            {
                "id": "fa0f4e46-7eab-4e5c-9d31-30489e69fe2e",
                "createdAt": "2022-11-11T18:38:46.266725Z",
                "updatedAt": "2022-11-11T18:38:46.266725Z",
                "name": "React",
                "code": "react"
            },
            {
                "id": "b403ef95-4269-4830-bbb6-8e56e5ec0af4",
                "createdAt": "2022-11-11T18:38:46.266725Z",
                "updatedAt": "2022-11-11T18:38:46.266725Z",
                "name": "Google Cloud",
                "code": "gcloud"
            }
        ]
    }
}
```

### CODE REVIEW

File: pkg/handler/project/project.go [hunk 1]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+// @Failure 400 {object} view.ErrorResponse
+// @Failure 404 {object} view.ErrorResponse
+// @Failure 500 {object} view.ErrorResponse
+// @Router /projects/{id}/general-info [put]
+func (h *handler) UpdateGeneralInfo(c *gin.Context) {
+	projectID := c.Param("id")
+
+	var body UpdateGeneralInfoInput
+	if err := c.ShouldBindJSON(&body); err != nil {
+		if err != nil {
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:926]: duplicate

File: pkg/handler/project/project.go [hunk 2]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+			c.JSON(http.StatusBadRequest, view.CreateResponse[any](nil, nil, err, body, ""))
+			return
+		}
+	}
+
+	// TODO: can we move this to middleware ?
+	l := h.logger.Fields(logger.Fields{
+		"handler": "project",
+		"method":  "UpdateGeneralInfo",
+		"request": body,
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:934]: need projectID

File: pkg/handler/project/project.go [hunk 3]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+	})
+
+	// Check project exists
+	check, err := h.store.Project.Exists(h.repo.DB(), projectID)
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:938]: `check` is a bit obscure, should replace by another name

File: pkg/handler/project/project.go (hunk 1 above)
Comment (datphamcode295) [reply pkg/handler/project/project.go:926]: updated

File: pkg/handler/project/project.go [hunk 4]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+	}
+
+	if !check {
+		l.Error(err, "error project not found")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, ErrProjectNotFound, body, ""))
+		return
+	}
+
+	// Check country exists
+	check, err = h.store.Country.Exists(h.repo.DB(), body.CountryID.String())
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:952]: same as above

File: pkg/handler/project/project.go [hunk 5]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+	check, err := h.store.Project.Exists(h.repo.DB(), projectID)
+	if err != nil {
+		l.Error(err, "error check existence of project")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, body, ""))
+		return
+	}
+
+	if !check {
+		l.Error(err, "error project not found")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, ErrProjectNotFound, body, ""))
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:947]: status code = 404

File: pkg/handler/project/project.go [hunk 6]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+	check, err = h.store.Country.Exists(h.repo.DB(), body.CountryID.String())
+	if err != nil {
+		l.Error(err, "error check existence of country")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, body, ""))
+		return
+	}
+
+	if !check {
+		l.Error(err, "error country not found")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, ErrCountryNotFound, body, ""))
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:961]: status code = 404

File: pkg/handler/project/project.go [hunk 7]
```diff
@@ -903,3 +903,121 @@ func (h *handler) Details(c *gin.Context) {
...
+		l.Error(err, "error when finding stacks")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, body, ""))
+		return
+	}
+
+	stackMap := model.ToStackMap(stacks)
+	for _, sID := range body.Stacks {
+		_, ok := stackMap[sID]
+		if !ok {
+			l.Error(errStackNotFound(sID.String()), "error stack not found")
```
Comment (nguyennh4522) [inline pkg/handler/project/project.go:977]: `error` in the error message doesn't make sense (same as with all error message)

File: pkg/handler/project/project_test.go [hunk 8]
```diff
@@ -508,3 +509,75 @@ func TestHandler_Details(t *testing.T) {
...
+			ctx.Params = gin.Params{gin.Param{Key: "id", Value: tt.id}}
+			ctx.Request = httptest.NewRequest("PUT", fmt.Sprintf("/api/v1/projects/%s/general-info", tt.id), bodyReader)
+			ctx.Request.Header.Set("Authorization", "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTkzMjExNDIsImlkIjoiMjY1NTgzMmUtZjAwOS00YjczLWE1MzUtNjRjM2EyMmU1NThmIiwiYXZhdGFyIjoiaHR0cHM6Ly9zMy1hcC1zb3V0aGVhc3QtMS5hbWF6b25hd3MuY29tL2ZvcnRyZXNzLWltYWdlcy81MTUzNTc0Njk1NjYzOTU1OTQ0LnBuZyIsImVtYWlsIjoidGhhbmhAZC5mb3VuZGF0aW9uIiwicGVybWlzc2lvbnMiOlsiZW1wbG95ZWVzLnJlYWQiXSwidXNlcl9pbmZvIjpudWxsfQ.GENGPEucSUrILN6tHDKxLMtj0M0REVMUPC7-XhDMpGM")
+			metadataHandler := New(storeMock, testRepoMock, serviceMock, loggerMock)
+
+			metadataHandler.UpdateGeneralInfo(ctx)
+			expRespRaw, err := ioutil.ReadFile(tt.wantResponsePath)
+			require.NoError(t, err)
+
+			require.JSONEq(t, string(expRespRaw), string(w.Body.Bytes()), "[Handler.UpdateProjectStatus] response mismatched")
```
Comment (nguyennh4522) [inline pkg/handler/project/project_test.go:580]: wrong message

File: pkg/handler/project/request.go [hunk 9]
```diff
@@ -15,6 +15,13 @@ type GetListProjectInput struct {
 	Type   string `form:"type" json:"type"`
 }
 
+type UpdateGeneralInfoInput struct {
+	Name      string       `form:"name" json:"name" binding:"required"`
+	StartDate *time.Time   `form:"startDate" json:"startDate"`
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/handler/project/request.go:20]: client just send date, not timestamp

File: pkg/handler/project/testdata/update_general_info/200.json [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    "data": {
+        "name": "Fortress",
+        "startDate": "1990-01-02T00:00:00Z",
+        "country": {
+            "id": "4ef64490-c906-4192-a7f9-d2221dadfe4c",
+            "createdAt": "2022-11-08T08:06:56.068148Z",
+            "updatedAt": "2022-11-08T08:06:56.068148Z",
+            "name": "Vietnam",
+            "code": "+84",
+            "cities": [
```
Comment (nguyennh4522) [inline pkg/handler/project/testdata/update_general_info/200.json:11]: don't send cities if client doesn't need it (ask client again)

File: pkg/store/project/project.go [hunk 11]
```diff
@@ -93,3 +96,18 @@ func (s *store) One(db *gorm.DB, id string) (*model.Project, error) {
 		First(&project).
 		Error
 }
+
+func (s *store) UpdateGeneralInfo(db *gorm.DB, body UpdateGeneralInfoInput, id string) (*model.Project, error) {
+	project := &model.Project{}
+	pp.Println(body)
```
Comment (nguyennh4522) [inline pkg/store/project/project.go:102]: remove log

Comment (datphamcode295) [reply pkg/store/project/project.go:102]: deleted

File: pkg/store/project/project.go [hunk 12]
```diff
@@ -71,12 +72,14 @@ func (s *store) Create(db *gorm.DB, project *model.Project) error {
 
 // Exists return true/false if project exist or not
 func (s *store) Exists(db *gorm.DB, id string) (bool, error) {
-	var record struct {
+	type res struct {
 		Result bool
 	}
-
+	fmt.Println(id)
```
Comment (namnhce) [inline pkg/store/project/project.go:77]: remove Println

**Review by namnhce — CHANGES_REQUESTED** [review]

**Review by namnhce — APPROVED** [review]

**Pull request merged**